packages/multichain/src/middlewares/MultichainMiddlewareManager.ts

```diff
@@ -0,0 +1,132 @@
+import type {
+  JsonRpcEngineEndCallback,
+  JsonRpcEngineNextCallback,
+} from '@metamask/json-rpc-engine';
+import type {
+  Json,
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+} from '@metamask/utils';
+
+import type { ExternalScopeString } from '../scope/types';
+
+export type ExtendedJsonRpcMiddleware = {
+  (
+    req: JsonRpcRequest & { scope: string },
+    res: PendingJsonRpcResponse<Json>,
+    next: JsonRpcEngineNextCallback,
+    end: JsonRpcEngineEndCallback,
+  ): void;
+  destroy?: () => void | Promise<void>;
+};
+
+type MiddlewareKey = {
+  scope: ExternalScopeString;
+  origin: string;
+  tabId?: number;
+};
+type MiddlewareEntry = MiddlewareKey & {
+  middleware: ExtendedJsonRpcMiddleware;
+};
+
+export class MultichainMiddlewareManager {
```

Comment:
What is the purpose of this class? Is it worth adding a blurb here?

Reply:
how about this https://github.com/MetaMask/core/pull/4813/commits/840d27da4ab9d3e72f437d3a5f03e81e51a7fbf3